sql/analyze.go

```diff
@@ -0,0 +1,928 @@
+			op = parser.EQ
+		case parser.GT:
+			op = parser.LT
+		case parser.GE:
+			op = parser.LE
+		case parser.LT:
+			op = parser.GT
+		case parser.LE:
+			op = parser.GE
+		case parser.Like:
+			op = parser.NotLike
+		case parser.NotLike:
+			op = parser.Like
+		case parser.SimilarTo:
+			op = parser.NotSimilarTo
+		case parser.NotSimilarTo:
+			op = parser.SimilarTo
+		default:
+			return parser.DBool(true)
+		}
+		return simplifyExpr(&parser.ComparisonExpr{
+			Operator: op,
+			Left:     t.Left,
+			Right:    t.Right,
+		})
+
+	case *parser.AndExpr:
+		// De Morgan's Law: (a AND b) -> (NOT a) OR (NOT b)
+		return simplifyExpr(&parser.OrExpr{
+			Left:  &parser.NotExpr{Expr: t.Left},
+			Right: &parser.NotExpr{Expr: t.Right},
+		})
+
+	case *parser.OrExpr:
+		// De Morgan's Law: NOT (a OR b) -> (NOT a) AND (NOT b)
+		return simplifyExpr(&parser.AndExpr{
+			Left:  &parser.NotExpr{Expr: t.Left},
+			Right: &parser.NotExpr{Expr: t.Right},
+		})
+	}
+	return parser.DBool(true)
+}
+
+func simplifyAndExpr(n *parser.AndExpr) parser.Expr {
+	// a AND b AND c AND d -> [a, b, c, d]
+	exprs := splitAndExpr(n, nil)
+	for i := range exprs {
+		exprs[i] = simplifyExpr(exprs[i])
+		if d, ok := exprs[i].(parser.DBool); ok && !bool(d) {
+			return d
+		}
+	}
+
+	// Loop over the expressions looking for simplifications.
+outer:
+	for i := len(exprs) - 1; i >= 0; i-- {
+		for j := i - 1; j >= 0; j-- {
+			exprs[j], exprs[i] = simplifyOneAndExpr(exprs[j], exprs[i])
+			if d, ok := exprs[j].(parser.DBool); ok && !bool(d) {
+				return d
+			}
+			if d, ok := exprs[i].(parser.DBool); ok && bool(d) {
+				exprs[i] = nil
+			}
+			if exprs[i] == nil {
+				// We found a simplification. Strip off the expression that is now nil
+				// and continue the outer loop.
+				exprs = exprs[:len(exprs)-1]
+				continue outer
+			}
+		}
+	}
+
+	// Reform the AND expressions.
+	if len(exprs) == 1 {
+		return exprs[0]
+	}
+	a := &parser.AndExpr{Left: exprs[len(exprs)-2], Right: exprs[len(exprs)-1]}
+	for i := len(exprs) - 3; i >= 0; i-- {
+		a = &parser.AndExpr{Left: exprs[i], Right: a}
+	}
+	return a
+}
+
+func simplifyOneAndExpr(left, right parser.Expr) (parser.Expr, parser.Expr) {
+	lcmp, ok := left.(*parser.ComparisonExpr)
+	if !ok {
+		return left, right
+	}
+	rcmp, ok := right.(*parser.ComparisonExpr)
+	if !ok {
+		return left, right
+	}
+	if !isDatum(lcmp.Right) || !isDatum(rcmp.Right) {
+		return parser.DBool(true), nil
+	}
+	if !varEqual(lcmp.Left, rcmp.Left) {
+		return left, right
+	}
+
+	if reflect.TypeOf(lcmp.Right) != reflect.TypeOf(rcmp.Right) {
+		switch lcmp.Operator {
+		case parser.EQ, parser.GT, parser.GE, parser.LT, parser.LE:
+			switch rcmp.Operator {
+			case parser.EQ, parser.GT, parser.GE, parser.LT, parser.LE:
+				return parser.DBool(false), nil
+			}
+		}
+		return left, right
+	}
+
+	ldatum := lcmp.Right.(parser.Datum)
+	rdatum := rcmp.Right.(parser.Datum)
+	cmp := ldatum.Compare(rdatum)
+
+	// TODO(pmattis): Figure out how to generate this logic.
```

Comment:
I wouldn't bother trying to review this switch or the one in `simplifyOneOrExpr`. Instead, look at the test cases in `TestSimplify{And,Or}Expr`.